[lldb-dap] Updating RequestHandler to encode/decode arguments and response.

ashgti opened the pull request:
This is a work in progress refactor to add explicit types instead of
generic 'llvm::json::Value' types to the DAP protocol.

This updates RequestHandler to have take the type of the arguments and
response body for serialization for requests.

The 'source' and 'disconnect' request is updated to show how the new flow
works and includes serialization handling for optional arguments and 'void' 
responses.

This is built on top of #130026

ashgti committed:
[lldb-dap] Updating RequestHandler to encode/decode arguments and response.

This is a work in progress refactor to add explicit types instead of
generic 'llvm::json::Value' types to the DAP protocol.

This updates RequestHandler to have take the type of the arguments and
response body for serialization for requests.

The 'source' request is updated to show how the new flow works.

vogelsgesang commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -57,235 +68,288 @@ class RequestHandler {
+    response.command = request.command;
+    Args arguments;
+    llvm::json::Path::Root root;
+    if (request.rawArguments &&
+        !fromJSON(request.rawArguments, arguments, root)) {
```
I think we should handle the case of missing arguments here...
Something like

```
   if constexpr (!std::is_same_v< Args, std::monostate>) {
      if (!request.rawArguments) {
          // Send error...
          return;
      }
      if (!fromJSON(request.rawArguments, arguments, root)) {
          // Send error...
          return;
      }
   }
```

ashgti replied:
Technically, they're optional, although all the existing request's require an argument today.

vogelsgesang replied:
> although all the existing request's require an argument today.

Afaict, the [BreakpointLocationsArguments](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_BreakpointLocations), [ConfigurationDone](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_ConfigurationDone), [Disconnect](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Disconnect), [RestartRequest](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Restart) and a couple more all have an optional `arguments` parameter.

Initially, I thought we would some commands which always take arguments and some commands which never accept arguments. Apparently, I was wrong. Instead, we apparently have commands which *optionally* accept arguments, and commands which always expect arguments, but no commands which never accept arguments.

I think we should produce a clear error message for commands which miss their required arguments.
Not sure how we want to model the distinction between optional and required arguments? Maybe something like the following distinction?

```
/// Requires `arguments` to be provided
/// Not providing `arguments` is an error
class MyRequestHandler : public RequestHandler<Arguments, ResponseBody> {

/// `arguments` are optional
class MyRequestHandler : public RequestHandler<std::optional<Arguments>, ResponseBody> {
```

ashgti replied:
I updated this to check if the `Args` are optional and fail the request if args are not provided.

I also added a `VoidResponse` helper that uses `std::monostate` to not return a response body.

This should help ensure we get arguments when we expect them and handle responses without a body.

LMKWYT

vogelsgesang commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -57,235 +68,288 @@ class RequestHandler {
+      llvm::raw_string_ostream OS(parseFailure);
+      root.printErrorContext(request.rawArguments, OS);
+      response.success = false;
+      response.message = parseFailure;
```
message is not meant to contain a user-readable string, It is meant to contain an enum-like value.

I know that large parts of DAP don't do this correctly. But for newly written code, I think we should directly be protocol-compliant. Or at least add a `FIXME` similar to how you did further down

We might want to extend our error message enum to also have an `protocolViolation` in addition to `cancelled` and `notStopped`. The more descriptive `parseFailure` string should go into the `message` inside the `ErrorResponse`.
I think this one we would like to mark as `showUser: true`?

ashgti replied:
I have a FIXME for addressing this, but to minimize the changes I kept this mostly backwards compatible with the current approach.

I can follow up this PR with a `DAPError` type for enhanced error reporting.

vogelsgesang commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -57,235 +68,288 @@ class RequestHandler {
+    }
+
+    auto ResponseBody = Run(arguments);
+    // FIXME: Add a dedicated DAPError for enhanced errors that are user
+    // visibile.
```
```suggestion
    // FIXME: Add a dedicated DAPError for enhanced errors that are
    // user-visible.
```

vogelsgesang commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -240,6 +240,137 @@ using Message = std::variant<Request, Response, Event>;
+  std::optional<int64_t> sourceReference;
+  std::optional<PresentationHint> presentationHint;
+
+  // unsupproted keys origin, sources, adapterData, checksums
```
```suggestion
  // unsupported keys: origin, sources, adapterData, checksums
```

ashgti replied:
Done.

vogelsgesang commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -240,6 +240,137 @@ using Message = std::variant<Request, Response, Event>;
+  // unsupproted keys origin, sources, adapterData, checksums
+};
+bool fromJSON(const llvm::json::Value &, Source &, llvm::json::Path);
+llvm::json::Value toJSON(const Source &);
```
do we actually need `toJSON`? `fromJSON` should be sufficient for `Source`?
From what I see, you didn't even implement `toJSON` for `Source`, so this is an unused declaration

```suggestion
```

vogelsgesang replied:
I see that we will need both in the future.

However, in your commit, the `toJSON` function is not even implemented, it is only declared. Calling it would lead to a linker error.

I see no benefit in introducing just the declaration in this commit. Either introduce both the declaration and implementation as part of this commit, or remove the declaration and introduce both declaration and implementation in a follow-up commit

ashgti replied:
Oh, I see what you mean, removed the `toJSON` since its not implemented.

vogelsgesang commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -240,6 +240,137 @@ using Message = std::variant<Request, Response, Event>;
+//   },
+//   "required": [ "sourceReference" ]
+// },
+struct SourceArguments {
```
I think we should change how we structure our files.

IMO, we should have `SourceArguments`, `SourceResponse` and `SourceRequestHandler` all in `Handlers/SourceRequestHandler.{h,cpp}`. That way, we have all the `SourceRequest`-related things in one place.

Not sure where `Source` should go. If it's only used by the `SourceRequestHandler`, I would also put it into `Handlers/SourceRequestHandler.{h,cpp}`, otherwise we might move it to `ProtocolTypes.{h,cpp}` or even `ProtocolTypes/Source.{h,cpp}`

vogelsgesang replied:
> otherwise we might move it to ProtocolTypes.{h,cpp} or even ProtocolTypes/Source.{h,cpp}

and maybe move the existing `Protocol.{h,cpp}` to `BaseProtocol.{h,cpp}`? To keep the base protocol separated from the other request / response / event types?

ashgti replied:
I personally like the protocol being isolated into its own namespace and its own file.

There are actually a few examples of this already in llvm, e.g. [clang-tools-extra/clangd/Protocol.h](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/Protocol.h) or [mlir/include/mlir/Tools/lsp-server-support/Protocol.h](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Tools/lsp-server-support/Protocol.h).

This is nice because we can easily look in the file for other types defined in the protocol and add new types as we need them. For example, you mentioned the `Source` type, which is used in a few events and requests. 

The [DAP spec](https://microsoft.github.io/debug-adapter-protocol/specification) already groups at a high level into `Base Protocol`, `Events`, `Requests`, `Reverse Requests` and `Types`, which I was roughly following in this file using `// MARK: <category>` headings between sections.

I only left the description of the `SourceRequest` next to the handler for reference for the implementation.

I'm fine with changing how we structure things, but I liked having the distinction between `lldb_dap::protocol` for the underlying protocol, `lldb_dap` for our intermediate components and `lldb`/`lldb_private` for the SB API layer of the debugger itself.

vogelsgesang replied:
> I'm fine with changing how we structure things, but I liked having the distinction between lldb_dap::protocol for the underlying protocol, lldb_dap for our intermediate components and lldb/lldb_private for the SB API layer of the debugger itself.

Indeed, you brought up a couple of good arguments for keeping the protocol separate. I am undecided. Let's see if anyone else jumps in and also has additional arguments in either direction. Otherwise, I would be fine with either way.

> The [DAP spec](https://microsoft.github.io/debug-adapter-protocol/specification) already groups at a high level into Base Protocol, Events, Requests, Reverse Requests and Types, which I was roughly following in this file using // MARK: <category> headings between sections.

Even if we don't move `*Args`, `*Request` and `*Response` types to the corresponding `*RequestHandler`, I think we should still separate the base protocol from the remaining protocol types. I would prefer separate files over `// MARK: <category>` comments for that purpose, i.e. a `Protocol.{h,cpp}` (containing the events, request and response types) and a separate `BaseProtocol.{h,cpp}` (containing the base protocol)

ashgti replied:
@JDevlieghere Any thoughts on the organization of these types?

I was keeping things simple by putting the types in a single place, but we don't strictly need to do that.

We could have `ProtocolBase.h`, `ProtocolTypes.h`, `ProtocolRequests.h`, `ProtocolEvents.h`, etc.

JDevlieghere replied:
I can see an argument for either. If it were up to me, I'd go with the different `Protocol*` files in a `Protocol` subdirectory. My (entirely subjective) reasoning is that the protocol should match the spec, rather than how it's used in our implementation and therefore it warrants standing on its own. Things were different when we didn't have the abstraction and type safety offered by the Protocol class and we had to consult the spec to decode a message, but now that's not a concern anymore if you're using an IDE or editor with code completion.

ashgti replied:
Split up the `Protocol.{h,cpp}` > `Protocol/Protocol{Base,Requests,Types}.{h,cpp}`.

vogelsgesang reviewed (commented):
👍 I like the overall direction this is going

JDevlieghere commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -240,6 +240,137 @@ using Message = std::variant<Request, Response, Event>;
+};
+bool fromJSON(const llvm::json::Value &, SourceArguments &, llvm::json::Path);
+
+// "SourceResponse": {
+//   "allOf": [ { "$ref": "#/definitions/Response" }, {
+//     "type": "object",
+//     "description": "Response to `source` request.",
+//     "properties": {
+//       "body": {
+//         "type": "object",
+//         "properties": {
+//           "content": {
+//             "type": "string",
+//             "description": "Content of the source reference."
+//           },
+//           "mimeType": {
+//             "type": "string",
+//             "description": "Content type (MIME type) of the source."
+//           }
+//         },
+//         "required": [ "content" ]
+//       }
+//     },
+//     "required": [ "body" ]
+//   }]
+// },
+struct SourceResponseBody {
+  std::string content;
+  std::optional<std::string> mimeType;
+};
```
I agree with the reasoning and I'm on board with removing it.

ashgti replied:
Removed the json schema and instead moved comments to the individual c++ reprs.

ashgti committed:
Adjusting comments to not be a copy-paste of the DAP json spec but instead extracting the comments to their c++ repr.

ashgti committed:
Updating fixme comment

vogelsgesang commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -30,216 +30,170 @@ namespace lldb_dap::protocol {
+/// Response for a request.
 struct Response {
   enum class Message {
+    /// the request was cancelled
     cancelled,
+    /// the request may be retried once the adapter is in a 'stopped' state
```
```suggestion
    /// The request was cancelled
    cancelled,
    /// The request may be retried once the adapter is in a 'stopped' state
```

ashgti committed:
Apply suggestions from code review

Co-authored-by: Adrian Vogelsgesang <adrian.vogelsgesang@tum.de>

ashgti commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -30,216 +30,170 @@ namespace lldb_dap::protocol {
   std::string command;
+
+  /// Object containing arguments for the command.
   std::optional<llvm::json::Value> rawArguments;
```
I used 'raw' as a prefix because we have the basic `json::Value` that we later make into the arguments type (e.g. when we convert this field into `SourceArguments`).

vogelsgesang replied:
For all other types, we have the same C++ member name like the JSON schema.

In Typescript this is defined as `any`, which (also in TypeScript) needs to be casted before being used. Nevertheless the authors of the Typescript types decided to name this `arguments` and not `rawArguments` in their schema.

That being said, I understand where you are coming from, and am not opposed to keeping it as is. If you prefer keeping `rawArguments`, we should point out in the comment, that we deviate from our normal naming convention here 

```
  /// Object containing arguments for the command.
  /// Serialized as `arguments` (not `rawArguments`) in JSON.
  std::optional<llvm::json::Value> rawArguments;
```

ashgti replied:
Added a comment and removed the `raw` prefix from these. I think your right that its not as clear that these map the the corresponding JSON fields.

ashgti committed:
Adding checks for missing arguments and added a `VoidResponse` for messages without a response body.

To validate both of these changes I also migrated the 'disconnect' request to have new types, since it has an `optional<DisconnectArguments>` and a `VoidResponse` body.

ashgti commented:
> What is your overall plan for this PR?
> 
> I assume the first step is to get buy-in from all reviewers into the new structure for our JSON types. But what comes after? Are you planning to update this PR to cover all JSON types / requests / responses? Are you planning to land this PR for the `Source*` types only (as currently in review) and do the remaining types in follow-up PRs?
> 
> (Both approaches would be fine by me. Just trying to understand the approach, so I know when it's time to hit the "Approve" button 😉)

I was thinking of doing this more incrementally. I was planning to build on this for handling the 'cancel' request first but then I'd like to follow up with requests to use the new explicit types for handling arguments and responses.

ashgti committed:
Moving the lldb::SBError > llvm::Error into LLDBUtils and making the `DAP::Disconnect` method a little more ergnomic.

ashgti committed:
Applying clang-format.

labath commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -57,235 +77,306 @@ class RequestHandler {
+class RequestHandler : public BaseRequestHandler {
+  using BaseRequestHandler::BaseRequestHandler;
+
+  void operator()(const llvm::json::Object &request) const override {
+    /* no-op, the other overload handles json coding. */
+  }
```
I think this wouldn't be necessary if you had something like a `LegacyRequestHandler` class which would declare the (abstract) function which takes the json object, and implements the protocol::Request version to forward to that). The base interface would only contain the protocol::Request overload.

labath reviewed (commented):
Just a small drive-by. This generally looks good to me, but I'll leave the approval to someone else.

vogelsgesang commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -248,6 +246,7 @@ struct DAP {
   // Serialize the JSON value into a string and send the JSON packet to
   // the "out" stream.
   void SendJSON(const llvm::json::Value &json);
+  void Send(const protocol::Message &message);
```
```suggestion
  /// Serialize the JSON value into a string and send the JSON packet to
  /// the "out" stream.
  void SendJSON(const llvm::json::Value &json);
  /// Send the given message to the client
  void Send(const protocol::Message &message);
```

vogelsgesang commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -663,58 +671,65 @@ void DAP::SetTarget(const lldb::SBTarget target) {
     return true;
   }
 
+  if (log)
+    *log << "Unsupported protocol message" << std::endl;
```
use new log macro?

vogelsgesang commented on lldb/tools/lldb-dap/Protocol.h:
```diff
@@ -30,216 +30,197 @@ namespace lldb_dap::protocol {
+  /// skipped on stepping.
+  std::optional<PresentationHint> presentationHint;
+
+  // unsupported keys origin, sources, adapterData, checksums
```
```suggestion
  // unsupported keys: origin, sources, adapterData, checksums
```

vogelsgesang commented on lldb/tools/lldb-dap/Handler/RequestHandler.cpp:
```diff
@@ -6,8 +6,9 @@
-#include "RequestHandler.h"
+#include "Handler/RequestHandler.h"
 #include "DAP.h"
+#include "Handler/ResponseHandler.h"
```
why do we need this now?

vogelsgesang replied:
ah, and this was transitively included via `DAP.h` previously. Makes sense

ashgti committed:
Adding a LegacyRequestHandler to simplify the BaseRequestHandler.

vogelsgesang reviewed (approved):
Looks good to me.

Please take another look at the two remaining open comments ([one](https://github.com/llvm/llvm-project/pull/130090#discussion_r1996417864), [two](https://github.com/llvm/llvm-project/pull/130090#discussion_r1991231163)) and then feel free to merge this

ashgti committed:
Adjusting the names of the "arguments" and "body" to remove the raw prefix. The types should be enough to tell that theyre not specific values and this helps with the naming consistency.

ashgti committed:
Splitting Protocol.{h,cpp} into Protocol/Protocol{Base,Types,Requests}.{h,cpp}.

ashgti committed:
Adding a newline to ProtocolRequests.h and adjusting whitespace in ProtocolBase.cpp.

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -663,58 +671,64 @@ void DAP::SetTarget(const lldb::SBTarget target) {
-      llvm::StringRef message = GetString(object, "message");
-      if (message.empty()) {
-        message = "Unknown error, response failed";
+      std::string message = "Unknown error, response failed";
+      if (resp->message) {
+        message = std::visit(
+            llvm::makeVisitor(
+                [](const std::string &message) -> std::string {
+                  return message;
+                },
+                [](const protocol::Response::Message &message) -> std::string {
```
Any reason these can't remain StringRefs? They're all string literals, right?

JDevlieghere commented on lldb/tools/lldb-dap/LLDBUtils.cpp:
```diff
@@ -162,4 +162,13 @@ GetEnvironmentFromArguments(const llvm::json::Object &arguments) {
   return envs;
 }
 
+llvm::Error takeError(const lldb::SBError &error) {
```
How about `ToError` to match `Status::ToError`? 
```suggestion
llvm::Error ToError(const lldb::SBError &error) {
```

ashgti replied:
Done.

JDevlieghere commented on lldb/tools/lldb-dap/Protocol/ProtocolBase.cpp:
```diff
@@ -1,12 +1,12 @@
-//===-- Protocol.cpp ------------------------------------------------------===//
+//===-- ProtocolBase.cpp --------------------------------------------------===//
```
Why not just Protocol? Do you think that's to confusing with the other Protocol classes?

ashgti replied:
I used the names of the categories from the spec for the files, https://microsoft.github.io/debug-adapter-protocol/specification if you look at the navigation on the left its grouped into 'Base Protocol', 'Events', 'Requests', 'Reverse Requests' and 'Types'.

JDevlieghere replied:
Oh right, I didn't make the connection. Sounds good.

JDevlieghere reviewed (approved):
LGTM modulo comments

ashgti committed:
Applying reviewer suggestions.

ashgti committed:
Merge branch 'main' into request-handler-types
